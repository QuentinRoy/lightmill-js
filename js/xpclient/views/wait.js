/*jslint nomen: true, browser:true*/
/*global define */

define(['spin', 'jquery'], function (Spinner, $) {
    "use strict";

    function WaitView(parentView) {
        this._parentView = parentView;
        this._spinView = $('<div class="full-parent"></div>');
        this._spinner = new Spinner(this._opts);
    }

    WaitView.prototype = {

        _opts: {
            lines: 11, // The number of lines to draw
            length: 20, // The length of each line
            width: 10, // The line thickness
            radius: 30, // The radius of the inner circle
            corners: 1, // Corner roundness (0..1)
            rotate: 0, // The rotation offset
            direction: 1, // 1: clockwise, -1: counterclockwise
            color: '#5C5C5C', // #rgb or #rrggbb or array of colors
            speed: 1, // Rounds per second
            trail: 60, // Afterglow percentage
            shadow: false, // Whether to render a shadow
            hwaccel: false, // Whether to use hardware acceleration
            className: 'spinner', // The CSS class to assign to the spinner
            zIndex: 2e9, // The z-index (defaults to 2000000000)
            top: 'auto', // Top position relative to parent in px
            left: 'auto' // Left position relative to parent in px
        },

        startWaiting: function () {
            this._spinView.appendTo(this._parentView);
            this._spinner.spin(this._spinView[0]);
        },
        stopWaiting: function () {
            this._spinner.stop();
            this._spinView.remove();
        }
    };

    return WaitView;

});