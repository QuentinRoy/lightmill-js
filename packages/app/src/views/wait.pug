.lightmill-base-app.lightmill-base-app-wait
  .lightmill-puck-wrapper
    .lightmill-puck.lightmill-puck-1
    .lightmill-puck.lightmill-puck-2
