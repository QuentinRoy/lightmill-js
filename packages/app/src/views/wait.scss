@import './common';

$puck-wrapper-size: 60px;
$puck-radius: 5px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .#{$class-prefix}puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    animation: $animation-name $rotation-duration infinite linear;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .#{$class-prefix}puck {
      position: absolute;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }

    .#{$class-prefix}puck-1 {
      top: 0;
      left: 0;
    }

    .#{$class-prefix}puck-2 {
      right: 0;
      bottom: 0;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
