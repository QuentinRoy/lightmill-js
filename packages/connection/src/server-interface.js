import fetch from 'unfetch';

/**
 * Check if the target experiment is loaded on the server.
 * @param  {Object}  serverInterface the server interface.
 * @param  {string}  experimentId    the id of the target experiment.
 * @return {Promise}                 true if the experiment is loaded on the server.
 */
export async function isExperimentLoadedOnServer(
  serverInterface,
  experimentId
) {
  const experiments = await serverInterface.experiments();
  return !!experiments[experimentId];
}

/**
 * Fetch the given experiment design xml file and post it to be imported on the server.
 * @param  {Object}  serverInterface      the server interface.
 * @param  {string}  experimentDesignAddr the address where to download the experiment design.
 * @return {Promise}                      resolves when the experiment has been imported.
 */
export async function importExperimentOnServer(
  serverInterface,
  experimentDesignAddr
) {
  const designReq = await fetch(experimentDesignAddr);
  const design = await designReq.text();
  return serverInterface.postExperimentDesign(design);
}

// Check if a given xhr type is a json type.
const isJsonType = type => type.endsWith('json');

// Fetch an address, expecting a json answer, and throwing if the requests failed.
const jsonThrowingFetch = (...args) =>
  fetch(...args).then(resp => {
    if (resp.ok) {
      // Every response are supposed to be json.
      return resp.json();
    }
    // If the requests has failed, throw an error.
    return resp.json().then(
      errorDescr => {
        const e = new Error(errorDescr.message);
        e.type = errorDescr.type;
        throw e;
      },
      () => {
        throw new Error(resp.statusText);
      }
    );
  });

// From a server address and the path to an API, return the api address (the main point of this
// function is to deal with the different form that can be taken by api path).
function getApiAddress(serverAddress, apiPath) {
  if (typeof apiPath === 'string') {
    return [serverAddress, apiPath].join('/');
  } else if (apiPath) {
    return [serverAddress, ...apiPath].join('/');
  }
  return serverAddress;
}

// Ask data to a server.
function askServer(address, options = {}) {
  // Create the request header (all requests are xhr).
  const headers = Object.assign(
    { Accept: 'application/json' },
    options.headers
  );
  // Create the options for fetch.
  const fetchOptions = Object.assign({ cache: 'no-cache' }, options, {
    headers
  });
  // Fetch.
  return jsonThrowingFetch(address, fetchOptions);
}

// Send data to a server.
function sendToServer(
  address,
  content,
  contentType = 'application/json',
  options = {}
) {
  // Set up the request headers, the content type and the request as xhr.
  const headers = Object.assign(
    {
      'Content-Type': contentType,
      Accept: 'application/json'
    },
    options.headers
  );
  // Automatically stringify non string content if the content type is json.
  const body =
    typeof content !== 'string' && isJsonType(contentType)
      ? JSON.stringify(content)
      : content;
  // Create the options for fetch.
  const fetchOptions = Object.assign({ method: 'POST', body }, options, {
    headers
  });
  // Fetch.
  return jsonThrowingFetch(address, fetchOptions);
}

/**
 * An interface toward the web xp server.
 * @param {String} serverAddress The address of the server.
 * @param {String} [apiPath='api'] The path toward the server API.
 * @constructor
 */
export default function ServerInterface(serverAddress, apiPath = 'api') {
  if (!(this instanceof ServerInterface)) {
    throw new Error('ServerInterface must be called with new');
  }

  // Init the api address once and for all.
  const apiAddress = getApiAddress(serverAddress, apiPath);
  // Get the address of an API endpoint on the server using either a simple
  // string, or an array.
  const getAddress = addrOrPathList =>
    (typeof addrOrPathList === 'string'
      ? [apiAddress, addrOrPathList]
      : [apiAddress, ...addrOrPathList]
    ).join('/');
  const ask = (addrOrPathList, ...args) =>
    askServer(getAddress(addrOrPathList), ...args);
  const send = (addrOrPathList, ...args) =>
    sendToServer(getAddress(addrOrPathList), ...args);

  /**
   * @return {Promise<Object, Object>} A promise that resolves with a
   * dictionary of the available experiments.
   */
  this.experiments = () => ask('experiments');

  /**
   * @param {String} experimentId  the experiment id.
   * @return {Promise<Object, Object>} A promise that resolves with the
   * description of an available run.
   */
  this.availableRun = experimentId =>
    ask(['experiment', experimentId, 'available_run']);

  /**
   * Request the basic information about a run.
   * @param  {String} experimentId The experiment id.
   * @param  {String} runId The run id.
   * @return {Promise} A promise that resolves with the
   * description of a run.
   */
  this.run = (experimentId, runId) => ask(['run', experimentId, runId]);

  /**
   * Request the plan of a run.
   * @param  {String} experimentId The experiment id.
   * @param  {String} runId The run id.
   * @return {Promise} A promise that resolves with the plan of
   * a run (including blocks and trials).
   */
  this.plan = (experimentId, runId) =>
    ask(['run', experimentId, runId, 'plan']);

  /**
   * Request a lock on a run.
   * @param  {String} experimentId The experiment id.
   * @param  {String} runId The run id.
   * @return {Promise} A promise that resolves with the run's lock.
   */
  this.lock = (experimentId, runId) =>
    ask(['run', experimentId, runId, 'lock']);

  /**
   * Request the current trial of a run (last non completed trial).
   * @param  {String} experimentId The experiment id.
   * @param  {String} runId The run id.
   * @return {Promise} A promise that resolves with the current (not yet
   * resolved) trial.
   */
  this.currentTrial = (experimentId, runId) =>
    ask(['run', experimentId, runId, 'current_trial']);

  /**
   * Post trial results (mark the current trial as completed on the server).
   * @param  {String} experimentId The experiment id.
   * @param  {String} runId The run id.
   * @param  {int} blockNumber The number of the block
   * @param  {int} trialNumber The number of the trial
   * @param  {Object} result The result of the trial.
   * @return {Promise} A promise that resolves one the trial results has been
   * recorded on the server.
   */
  this.postTrialResults = (
    experimentId,
    runId,
    blockNumber,
    trialNumber,
    result
  ) => send(['trial', experimentId, runId, blockNumber, trialNumber], result);
  this.postExperimentDesign = expeDesign =>
    send('import', expeDesign, 'text/xml');

  /**
   * Check if the target experiment is loaded on the server.
   * @param  {string}  experimentId the id of the target experiment.
   * @return {Promise} A promise that resolves with true if the experiment is
   * loaded on the server.
   */
  this.isExperimentLoadedOnServer = experimentId =>
    isExperimentLoadedOnServer(this, experimentId);

  /**
   * Fetch the given experiment design xml file and post it to be imported on
   * the server.
   * @param  {string}  experimentDesignAddr the address where to download the
   * experiment design.
   * @return {Promise} A promise that resolves when the experiment has been
   * imported.
   */
  this.importExperimentOnServer = experimentDesignAddr =>
    importExperimentOnServer(this, experimentDesignAddr);
}
