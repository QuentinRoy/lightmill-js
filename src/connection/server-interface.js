import fetch from 'unfetch';

const isJsonType = type => type.endsWith('json');

// Fetch an address, expecting a json answer, and throwing if the requests failed.
const jsonThrowingFetch = (...args) =>
  fetch(...args).then((resp) => {
    if (resp.ok) {
      // Every response are supposed to be json.
      return resp.json();
    }
    // If the requests has failed, throw an error.
    return resp.json().then(
      (errorDescr) => {
        const e = new Error(errorDescr.message);
        e.type = errorDescr.type;
        throw e;
      },
      () => {
        throw new Error(resp.statusText);
      }
    );
  });


// From a server address and the path to an API, return the api address (the main point of this
// function is to deal with the different form that can be tacken by api path).
const getApiAddress = (serverAddress, apiPath) => {
  if (typeof apiPath === 'string') {
    return [serverAddress, apiPath].join('/');
  } else if (apiPath) {
    return [serverAddress, ...apiPath].join('/');
  }
  return serverAddress;
};

// Returns an interface toward the provided server address.
export default function getServerInterface(serverAddress, apiPath = 'api') {
  // Init the api address once and for all.
  const apiAddress = getApiAddress(serverAddress, apiPath);

  // Get the address of an API endpoint on the server using either a simple string, or an
  // array.
  const getAddress = addrOrPathList =>
    (typeof addrOrPathList === 'string'
      ? [apiAddress, addrOrPathList]
      : [apiAddress, ...addrOrPathList]).join('/');

  // Ask data to the server.
  const ask = (addrOrPathList, options = {}) => {
    // Create the request header (all requests are xhr).
    const headers = Object.assign({ 'X-Requested-With': 'XMLHttpRequest' }, options.headers);
    // Create the options for fetch.
    const fetchOptions = Object.assign({ cache: 'no-cache' }, options, { headers });
    // Fetch.
    return jsonThrowingFetch(getAddress(addrOrPathList), fetchOptions);
  };

  // Send data to the server.
  const send = (addrOrPathList, content, contentType = 'application/json', options = {}) => {
    // Set up the request headers, the content type and the request as xhr.
    const headers = Object.assign(
      {
        'Content-Type': contentType,
        'X-Requested-With': 'XMLHttpRequest'
      },
      options.headers
    );
    // Automatically stringify non string content if the content type is json.
    const body =
      typeof content !== 'string' && (isJsonType(contentType) ? JSON.stringify(content) : content);
    // Create the options for fetch.
    const fetchOptions = Object.assign({ method: 'POST', body }, options, { headers });
    // Fetch.
    return jsonThrowingFetch(getAddress(addrOrPathList), fetchOptions);
  };

  return {
    experiments: () => ask('experiments'),
    availableRun: xpId => ask(['experiment', xpId, 'available_run']),
    run: (xpId, runId) => ask(['run', xpId, runId]),
    plan: (xpId, runId) => ask(['run', xpId, runId, 'plan']),
    lock: (xpId, runId) => ask(['run', xpId, runId, 'lock']),
    currentTrial: (xpId, runId) => ask(['run', xpId, runId, 'current_trial']),
    postTrialResults: (experimentId, runId, blockNumber, trialNumber, data) =>
      send(['trial', experimentId, runId, blockNumber, trialNumber], data),
    postExperimentDesign: expeDesign => send('import', expeDesign, 'text/xml')
  };
}
