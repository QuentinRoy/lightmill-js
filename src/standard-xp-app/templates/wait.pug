.xp-app-base.xp-app-base-wait
  .puck-wrapper
    .puck
    
