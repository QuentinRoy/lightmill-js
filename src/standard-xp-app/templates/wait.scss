@import './common';

$puck-wrapper-size: 400px;
$puck-radius: 25px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .puck {
      position: absolute;
      top: 0;
      left: 0;
      animation: $animation-name $rotation-duration infinite linear;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
