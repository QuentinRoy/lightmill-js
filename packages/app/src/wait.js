/* eslint-disable no-param-reassign */

import waitTemplate from './views/wait.pug';
import './views/wait.scss';

export default (node, message) => {
  node.innerHTML = waitTemplate({ message });

  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
