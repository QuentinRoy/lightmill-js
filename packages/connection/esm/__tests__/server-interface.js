import wait from 'wait-then';
import { isExperimentLoadedOnServer } from '../server-interface';

describe('isExperimentLoadedOnServer', async () => {
  it('checks if the experiment is available on the server', async () => {
    // Mock the server interface.
    const server = {
      experiments: jest.fn(() => wait(0).then(() => ({ xp1: {}, xp2: {} })))
    };
    expect(await isExperimentLoadedOnServer(server, 'xp1')).toBe(true);
    expect(server.experiments).toMatchSnapshot();
    expect(await isExperimentLoadedOnServer(server, 'not-here')).toBe(false);
    expect(server.experiments).toMatchSnapshot();
  });
});
