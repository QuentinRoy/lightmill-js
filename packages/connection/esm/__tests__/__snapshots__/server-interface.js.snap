// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isExperimentLoadedOnServer checks if the experiment is available on the server 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
}
`;

exports[`isExperimentLoadedOnServer checks if the experiment is available on the server 2`] = `
[MockFunction] {
  "calls": Array [
    Array [],
    Array [],
  ],
}
`;
